File: scripts/build-lesson-counts.js
#!/usr/bin/env node
/**
 * build-lesson-counts.js
 * Scans professor HTML files and counts actual lesson cards (id="ex-lesson-N").
 * Outputs /public/data/lesson-counts.json
 * Run: node scripts/build-lesson-counts.js
 */

const fs = require('fs');
const path = require('path');

const PROF_DIR = path.join(__dirname, '..', 'public', 'professor');
const OUT_FILE = path.join(__dirname, '..', 'public', 'data', 'lesson-counts.json');

// Skip non-main files (but track standalone aula files separately)
const SKIP_PATTERNS = [
    /-aula\d/,
    /-backup/,
    /-test/,
    /-v-[ab]/,
    /V2\./,
    /-new-?aula/,
    /-new\./,
    /-listening\./,
    /-palestra\./,
    /-speech-training\./,
    /^helen-mendes\./,  // aluna MOCK (template modelo) — fora do controle
];

const STANDALONE_AULA_RE = /^(.+)-aula(\d+)\.html$/;

function isMainFile(filename) {
    return !SKIP_PATTERNS.some(p => p.test(filename));
}

function countLessons(html) {
    const matches = html.match(/id="ex-lesson-/g);
    return matches ? matches.length : 0;
}

function extractName(html, slug) {
    // Try to get name from <h1> in header
    const h1Match = html.match(/<h1[^>]*>([^<]+)<\/h1>/);
    if (h1Match) return h1Match[1].trim();
    // Fallback: humanize slug
    return slug.split('-').map(w => w.charAt(0).toUpperCase() + w.slice(1)).join(' ');
}

function getStandaloneAulaNumbers(allFiles) {
    // Get standalone aula numbers (-aula2.html, -aula3.html, etc.) per slug
    const aulaNumbers = {};
    for (const file of allFiles) {
        const match = file.match(STANDALONE_AULA_RE);
        if (match) {
            const slug = match[1];
            const num = parseInt(match[2]);
            if (!aulaNumbers[slug]) aulaNumbers[slug] = [];
            aulaNumbers[slug].push(num);
        }
    }
    return aulaNumbers;
}

function getMaxLessonInMain(html) {
    // Find the highest lesson number in id="ex-lesson-N"
    const matches = html.match(/id="ex-lesson-(\d+)"/g);
    if (!matches) return 0;
    let max = 0;
    for (const m of matches) {
        const num = parseInt(m.match(/(\d+)/)[1]);
        if (num > max) max = num;
    }
    return max;
}

function run() {
    const allHtmlFiles = fs.readdirSync(PROF_DIR).filter(f => f.endsWith('.html'));
    const mainFiles = allHtmlFiles.filter(f => isMainFile(f));
    const standaloneAulaNumbers = getStandaloneAulaNumbers(allHtmlFiles);
    const results = [];

    for (const file of mainFiles) {
        const slug = file.replace('.html', '');
        const html = fs.readFileSync(path.join(PROF_DIR, file), 'utf-8');
        const lessonCount = countLessons(html);
        const maxInMain = getMaxLessonInMain(html);

        // Only count standalone aulas with numbers BEYOND what's in the main file
        const standaloneNums = standaloneAulaNumbers[slug] || [];
        const extraStandalone = standaloneNums.filter(n => n > maxInMain).length;
        const totalCount = lessonCount + extraStandalone;

        if (totalCount === 0) continue; // skip empty/redirect files

        const name = extractName(html, slug);

        results.push({
            slug,
            name,
            aulasHtml: totalCount
        });
    }

    results.sort((a, b) => a.name.localeCompare(b.name, 'pt-BR'));

    const output = {
        generatedAt: new Date().toISOString(),
        students: results
    };

    fs.mkdirSync(path.dirname(OUT_FILE), { recursive: true });
    fs.writeFileSync(OUT_FILE, JSON.stringify(output, null, 2));
    console.log(`Generated ${OUT_FILE} with ${results.length} students`);
}

run();

File: scripts/build_lesson_counts.py
#!/usr/bin/env python3
"""Porte Python do build-lesson-counts.js (node não existe no ambiente de dev).

Varre public/professor/*.html e conta as aulas REAIS por aluno:
  total = nº de id="ex-lesson-" no hub + standalones {slug}-aulaN.html com N
          acima do maior ex-lesson-N do hub.
Gera public/data/lesson-counts.json (mesmo formato do script node).

Uso: python3 scripts/build_lesson_counts.py
"""
import json
import re
import sys
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
PROF_DIR = ROOT / 'public' / 'professor'
OUT_FILE = ROOT / 'public' / 'data' / 'lesson-counts.json'

SKIP_PATTERNS = [
    r'-aula\d', r'-backup', r'-test', r'-v-[ab]', r'V2\.', r'-new-?aula',
    r'-new\.', r'-listening\.', r'-palestra\.', r'-speech-training\.',
    r'^helen-mendes\.',  # aluna MOCK (modelo de template) — não entra no controle
]
STANDALONE_AULA_RE = re.compile(r'^(.+)-aula(\d+)\.html$')


def is_main_file(name):
    return not any(re.search(p, name) for p in SKIP_PATTERNS)


def extract_name(html, slug):
    m = re.search(r'<h1[^>]*>([^<]+)</h1>', html)
    if m:
        return m.group(1).strip()
    return ' '.join(w.capitalize() for w in slug.split('-'))


def main():
    all_files = sorted(f.name for f in PROF_DIR.glob('*.html'))
    standalone = {}
    for f in all_files:
        m = STANDALONE_AULA_RE.match(f)
        if m:
            standalone.setdefault(m.group(1), []).append(int(m.group(2)))

    results = []
    for fname in all_files:
        if not is_main_file(fname):
            continue
        slug = fname[:-5]
        html = (PROF_DIR / fname).read_text(encoding='utf-8', errors='replace')
        lesson_count = len(re.findall(r'id="ex-lesson-', html))
        nums_in_main = [int(n) for n in re.findall(r'id="ex-lesson-(\d+)"', html)]
        max_in_main = max(nums_in_main, default=0)
        extra = [n for n in standalone.get(slug, []) if n > max_in_main]
        total = lesson_count + len(extra)
        if total == 0:
            continue
        results.append({
            'slug': slug,
            'name': extract_name(html, slug),
            'aulasHtml': total,
        })

    results.sort(key=lambda r: r['name'].lower())
    out = {
        'generatedAt': datetime.now(timezone.utc).isoformat(timespec='milliseconds').replace('+00:00', 'Z'),
        'students': results,
    }
    OUT_FILE.parent.mkdir(parents=True, exist_ok=True)
    OUT_FILE.write_text(json.dumps(out, indent=2, ensure_ascii=False) + '\n', encoding='utf-8')
    print(f'Generated {OUT_FILE.relative_to(ROOT)} with {len(results)} students')
    return 0


if __name__ == '__main__':
    sys.exit(main())
